File: codegen/features/xsls/backendless-codegen-xsls/ios/backendless-codegen/ObjC/FileService/FileService/BrowseViewController.m
#import "BrowseViewController.h"
#import "ImagePreviewViewController.h"
#import "FileObject.h"
#import <Backendless-Swift.h>

@interface BrowseViewController () {
    NSArray *mainData;
    DataStoreFactory *filesDataStore;
}
@end

@implementation BrowseViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    filesDataStore = [Backendless.shared.data of:[FileObject class]];
}

- (void)viewWillAppear:(BOOL)animated {
    [super viewWillAppear:animated];
    [self getAllEntitysAsync];
}

- (void)getAllEntitysAsync {
    [filesDataStore findWithResponseHandler:^(NSArray *photos) {
        if (photos.count == 0) {
            [self presentViewController:[UIAlertController alertControllerWithTitle:@"No files found" message:@"Please take some photos before you can browse them" preferredStyle:UIAlertControllerStyleAlert] animated:YES completion:^{
                [NSTimer scheduledTimerWithTimeInterval:3 repeats:NO block:^(NSTimer *timer) {
                    [self dismissViewControllerAnimated:YES completion:nil];
                }];
            }];
            [self.navigationController popToRootViewControllerAnimated:YES];
        }
        else {
            self->mainData = photos;
            [self.mainTableView reloadData];
        }
    } errorHandler:^(Fault *fault) {
        [self presentViewController:[UIAlertController alertControllerWithTitle:@"Error" message:fault.message preferredStyle:UIAlertControllerStyleAlert] animated:YES completion:^{
            [NSTimer scheduledTimerWithTimeInterval:3 repeats:NO block:^(NSTimer *timer) {
                [self dismissViewControllerAnimated:YES completion:nil];
            }];
        }];
        [self.navigationController popToRootViewControllerAnimated:YES];
    }];
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    return (mainData.count / 4) + ((mainData.count % 4) ? 1 : 0);
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    static NSString *cellIdentifier = @"Cell";
    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:cellIdentifier];
    for (int i = 1; i <= 4; i++) {
        if (indexPath.row * 4 + i > mainData.count)
            break;
        UIButton *button = (UIButton *)[cell viewWithTag:i];
        NSString *str = (NSString *)[[mainData objectAtIndex:indexPath.row * 4 + i - 1] path];
        NSURL *url = [NSURL URLWithString:str];
        button.hidden = NO;
        button.enabled = NO;
        NSURLSessionDownloadTask *downloadPhotoTask = [[NSURLSession sharedSession] downloadTaskWithURL:url completionHandler:^(NSURL *location, NSURLResponse *response, NSError *error) {
            UIImage *downloadedImage = [UIImage imageWithData: [NSData dataWithContentsOfURL:location]];
            dispatch_async(dispatch_get_main_queue(), ^{
                [button setImage:downloadedImage forState:UIControlStateNormal];
                button.enabled = YES;
            });
        }];
        [downloadPhotoTask resume];
    }
    return cell;
}

- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender {
    NSIndexPath *indexPath = [self.mainTableView indexPathForCell:(UITableViewCell *)[[sender superview] superview]];
    ImagePreviewViewController *imagePreview = (ImagePreviewViewController *)[segue destinationViewController];
    [imagePreview setMainImage:[(UIButton *)sender imageForState:UIControlStateNormal]];
    [imagePreview setIsUpload:YES];
    [imagePreview setFile:[mainData objectAtIndex:indexPath.row * 4 + [sender tag] - 1]];
}

- (IBAction)pressedRemoveAll:(id)sender {
    [Backendless.shared.file removeWithPath:@"img" responseHandler:^(NSInteger removed) {
        for (FileObject *file in self->mainData) {
            [self->filesDataStore removeWithEntity:file responseHandler:^(NSInteger removed) {
                [self presentViewController:[UIAlertController alertControllerWithTitle:@"Files removed" message:@"All files have been removed" preferredStyle:UIAlertControllerStyleAlert] animated:YES completion:^{
                    [NSTimer scheduledTimerWithTimeInterval:3 repeats:NO block:^(NSTimer *timer) {
                        [self dismissViewControllerAnimated:YES completion:nil];
                    }];
                }];
            } errorHandler:^(Fault *fault) {
                [self presentViewController:[UIAlertController alertControllerWithTitle:@"Error" message:fault.message preferredStyle:UIAlertControllerStyleAlert] animated:YES completion:^{
                    [NSTimer scheduledTimerWithTimeInterval:3 repeats:NO block:^(NSTimer *timer) {
                        [self dismissViewControllerAnimated:YES completion:nil];
                    }];
                }];
            }];
        }
    } errorHandler:^(Fault *fault) {
        [self presentViewController:[UIAlertController alertControllerWithTitle:@"Error" message:fault.message preferredStyle:UIAlertControllerStyleAlert] animated:YES completion:^{
            [NSTimer scheduledTimerWithTimeInterval:3 repeats:NO block:^(NSTimer *timer) {
                [self dismissViewControllerAnimated:YES completion:nil];
            }];
        }];
    }];
    [self.navigationController popToRootViewControllerAnimated:YES];
}

@end

File: codegen/features/xsls/backendless-codegen-xsls/ios/backendless-codegen/Swift/FileService/FileService/BrowseViewController.swift

import UIKit
import Backendless

class BrowseViewController: UIViewController, UITableViewDataSource, UITableViewDelegate {
    
    @IBOutlet weak var mainTableView: UITableView!
    
    var mainData: [FileObject]?
    
    override func viewDidLoad() {
        super.viewDidLoad()
        getAllEntitiesAsync()
    }
    
    func getAllEntitiesAsync() {
        Backendless.shared.data.of(FileObject.self).find(responseHandler: { photos in
            if photos.count == 0 {
                self.present(UIAlertController.init(title: "No files found", message: "Please take some photos before you can browse them", preferredStyle: .alert), animated: true, completion: {
                    Timer.scheduledTimer(withTimeInterval: 3, repeats:false, block: {_ in
                        self.dismiss(animated: true, completion: nil)
                    })
                })
                self.navigationController?.popToRootViewController(animated: true)
            }
            else {
                self.mainData = photos as? [FileObject]
                self.mainTableView.reloadData()
            }
        }, errorHandler: { fault in
            self.present(UIAlertController.init(title: "Error", message: fault.message, preferredStyle: .alert), animated: true, completion: {
                Timer.scheduledTimer(withTimeInterval: 3, repeats:false, block: {_ in
                    self.dismiss(animated: true, completion: nil)
                })
            })
            self.navigationController?.popToRootViewController(animated: true)
        })
    }
    
    @IBAction func removeAll(_ sender: Any) {
        Backendless.shared.file.remove(path: "img", responseHandler: { _ in
            for file in self.mainData! {
                Backendless.shared.data.of(FileObject.self).remove(entity: file, responseHandler: { removed in
                    self.present(UIAlertController.init(title: "Files removed", message: "All files have been removed", preferredStyle: .alert), animated: true, completion: {
                        Timer.scheduledTimer(withTimeInterval: 3, repeats:false, block: {_ in
                            self.dismiss(animated: true, completion: nil)
                        })
                    })
                    self.navigationController?.popToRootViewController(animated: true)
                }, errorHandler: { fault in
                    self.present(UIAlertController.init(title: "Error", message: fault.message, preferredStyle: .alert), animated: true, completion: {
                        Timer.scheduledTimer(withTimeInterval: 3, repeats:false, block: {_ in
                            self.dismiss(animated: true, completion: nil)
                        })
                    })
                    self.navigationController?.popToRootViewController(animated: true)
                })
            }
        }, errorHandler: { fault in
            self.present(UIAlertController.init(title: "Error", message: fault.message, preferredStyle: .alert), animated: true, completion: {
                Timer.scheduledTimer(withTimeInterval: 3, repeats:false, block: {_ in
                    self.dismiss(animated: true, completion: nil)
                })
            })
            self.navigationController?.popToRootViewController(animated: true)
        })
    }
    
    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        if mainData != nil {
            if (mainData!.count % 4 == 0) {
                return (mainData!.count / 4)
            }
            else {
                return (mainData!.count / 4 + 1)
            }
        }
        else {
            return 0
        }
    }
    
    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let cellIdentifier = "Cell"
        let cell = tableView.dequeueReusableCell(withIdentifier: cellIdentifier)
        for i in 1...4 {
            if (indexPath.row * 4 + i > (mainData?.count)!) {
                break
            }
            let button = cell?.viewWithTag(i) as! UIButton
            if let str = mainData?[indexPath.row * 4 + i - 1].path,
               let url = URL(string: str) {
                button.isHidden = false
                button.isEnabled = false
                                
                let downloadPhotoTask = URLSession.shared.downloadTask(with: url) { location, response, error in
                    if location != nil,
                       let data = try? Data(contentsOf: location!),
                       let downloadedImage = UIImage(data: data) {
                        DispatchQueue.main.async {
                            button.setImage(downloadedImage, for: .normal)
                            button.isEnabled = true
                        }
                    }
                }
                downloadPhotoTask.resume()
                
                /*NSURLSessionDownloadTask *downloadPhotoTask = [[NSURLSession sharedSession] downloadTaskWithURL:url completionHandler:^(NSURL *location, NSURLResponse *response, NSError *error) {
                 UIImage *downloadedImage = [UIImage imageWithData: [NSData dataWithContentsOfURL:location]];
                 dispatch_async(dispatch_get_main_queue(), ^{
                     [button setImage:downloadedImage forState:UIControlStateNormal];
                     button.enabled = YES;
                 });
             }];
             [downloadPhotoTask resume];*/
                
                
                
                
//                NSURLConnection.sendAsynchronousRequest(NSURLRequest.init(url: url! as URL) as URLRequest, queue: OperationQueue.main, completionHandler: { (response: URLResponse?, data: Data?, error: Error?) -> () in
//                    let responseUrl = response?.copy() as! HTTPURLResponse
//                    let statusCode = responseUrl.statusCode
//                    if (statusCode == 200) {
//                        button.setImage(UIImage.init(data: data!), for: .normal)
//                        button.isEnabled = true
//                    }
//                })
            }
        }
        return cell!
    }
    
    override func prepare(for segue: UIStoryboardSegue, sender: Any?) {
        if let imagePreview = segue.destination as? ImagePreviewViewController {
            imagePreview.mainImage = (sender as! UIButton).image(for: .normal)
            imagePreview.isUpload = true
            let cell = (sender as! UIButton).superview?.superview as! UITableViewCell
            let row = mainTableView.indexPath(for: cell)!.row
            imagePreview.file = mainData?[row * 4 + ((sender as AnyObject).tag)! - 1]
        }
    }
}
